flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java

```diff
@@ -0,0 +1,234 @@
+		Mockito.verify(miniCluster).triggerSavepoint(jobGraph.getJobID(), savepointPath, false);
+
+		jobClient.stopWithSavepoint(true, savepointPath);
+		Mockito.verify(miniCluster).stopWithSavepoint(jobGraph.getJobID(), savepointPath, true);
+	}
+
+	@Test
+	public void testSubmissionError() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+
+		// JobGraph is not a valid job
+		JobGraph jobGraph = new JobGraph();
+
+		try {
+			perJobMiniCluster.submitJob(jobGraph).get();
+		} catch (ExecutionException e) {
+			// That's what we want
+			assertThat(e.getMessage(), containsString("Error submitting job to MiniCluster"));
+		}
+
+		assertThatMiniClusterIsShutdown();
+	}
+
+	@Test
+	public void testMultipleExecutions() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+		{
+			JobClient jobClient = perJobMiniCluster.submitJob(getNoopJobGraph()).get();
+			jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
+			assertThatMiniClusterIsShutdown();
+		}
+		{
+			JobClient jobClient = perJobMiniCluster.submitJob(getNoopJobGraph()).get();
+			jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
+			assertThatMiniClusterIsShutdown();
+		}
+	}
+
+	@Test
+	public void testJobClientInteractionAfterShutdown() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+		JobClient jobClient = perJobMiniCluster.submitJob(getNoopJobGraph()).get();
+		jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
+		assertThatMiniClusterIsShutdown();
+
+		try {
+			jobClient.cancel();
+		} catch (IllegalStateException e) {
+			// That's what we want
+			assertThat(e.getMessage(), is("MiniCluster is not yet running or has already been shut down."));
+		}
+	}
+
+	private PerJobMiniCluster initializeMiniCluster() {
+		return new PerJobMiniCluster(new Configuration(), config -> {
+			miniCluster = new MiniCluster(config);
+			return miniCluster;
+		});
+	}
+
+	private void assertThatMiniClusterIsShutdown() {
+		assertThat(miniCluster.isRunning(), is(false));
+	}
+
+	private static JobGraph getNoopJobGraph() {
+		JobGraph jobGraph = new JobGraph();
+		JobVertex jobVertex = new JobVertex("jobVertex");
+		jobVertex.setInvokableClass(NoOpInvokable.class);
+		jobGraph.addVertex(jobVertex);
+		return jobGraph;
+	}
+
+	private static JobGraph getCancellableJobGraph() {
+		JobGraph jobGraph = new JobGraph();
+		JobVertex jobVertex = new JobVertex("jobVertex");
+		jobVertex.setInvokableClass(MyCancellableInvokable.class);
+		jobGraph.addVertex(jobVertex);
+		return jobGraph;
+	}
+
+	/**
+	 * Invokable which waits until it is cancelled.
+	 */
+	public static class MyCancellableInvokable extends CancelableInvokable {
+
+		private final Object lock = new Object();
+
+		public MyCancellableInvokable(Environment environment) {
+			super(environment);
+		}
+
+		@Override
+		public void invoke() throws Exception {
+			synchronized (lock) {
+				lock.wait();
```

Comment:
According to the JavaDocs of `wait`

>As in the one argument version, interrupts and spurious wakeups are possible, and this method should always be used in a loop:

Hence I would suggest to harden this usage.